### ISSUE TITLE
Unable to copy full wrapped word/line with double/triple click

### ISSUE THREAD

**Issue by tuv7041** [tmux/tmux#2414]
$TERM: rxvt-unicode-256color
Platform: debian testing (bullseye)
Tmux version: tmux 3.1b

Issue: When I try to copy either a word or a line that is wrapped, Tmux only selects a single line. It doesn't wrap around. I have a config that uses select-word and select-line. Using the default terminal, double click selects the full word, even if its wrapped. Triple click does the same but with the full line. Is there any way to make these work with multi-line wrapped text?

Thanks!

**Comment by nicm** [comment]
Double-client and triple-click select do not wrap at the moment, although I don't think this is particularly intentional.

**Comment by tuv7041** [comment]
Is there any way in which this behaviour can be implemented? (maybe a custom function in .tmux.conf?) I think it would be a great thing to have since it mimics the default behaviour of the copy and paste (but allows you to keep all the other tmux's great features.

**Comment by tuv7041** [comment]
Could it be done with a .tmux.conf function that selects the string until a newline is detected? I'm not sure if the the capability to scan for a certain character (or maybe a regex) exists on that level or would it need a source code change like you said.

**Comment by nicm** [comment]
It would be very complicated to do it without code changes

**Comment by nicm** [comment]
The problem here is that `select-word` uses `next-word-end` which stops at the edge of the screen as well as the actual end of the line. Likewise `select-line` uses `end-of-line`.

This is how most of the commands work (`end-of-line`, `start-of-line`, `next-word`/`next-word-end`, `copy-end-of-line` and so on).

I think it would be better if they worked like emacs and always used the actual end of line rather than the screen edge. I looked at changing them but it is complicated and I don't have time to do it now. Copy mode could do with quite a lot of tidying up.

So I will put it on the todo list and perhaps look at it in future (unless someone else is interested in doing it - @topcat001 did quite a bit of work on copy mode, maybe he is interested?).

**Comment by nicm** [comment]
I think it would be best if there was a set of helper functions that worked on the grid and all commands used those to do what they want, instead of the situation which we have now where one command can fire another and so on, which makes it hard to change things. But this is a lot of work to clean up I expect.

**Comment by topcat001** [comment]
I am aware of this problem and it bothers me as well. I have also been thinking about refactoring some of the copy-mode functions to make further work easier. It's a bit of work. I'll take a look when I'm able, thanks!

**Comment by nicm** [comment]
I was thinking about this and what I would probably try to end up with is an API that works on a grid (or possible a screen?) for navigating/querying it, a bit like the screen_write API for writing into it. So a struct like grid_reader which contains a virtual cursor position and functions like:

- void grid_reader_start(struct grid_reader *, struct grid *, u_int cx, u_int cy); - set up
- void grid_reader_get_cursor(struct grid_reader *, u_int *cx, u_int *cy);
- u_int grid_reader_line_length(struct grid_reader *); - length of current line
- grid_reader_cursor_forward - move cursor forward, accounting for wrapped lines, maybe with a flag to say if wrapping is allowed or not
- grid_reader_end_of_line - move to end of current line, accounting for wrapped lines
- grid_reader_next_word - move to next word

And so on, initially it could be pretty simple but eventually extend to searching and stringifying and all the other stuff. Some of the code is there already in grid.c, screen-write.c, window-copy.c.

Then we could build all the window-copy.c functions on this API and they would all work the same way and not have complicated interdependencies like end. Eg if we wanted to move to the next word + one char we could do something like:
~~~~
struct grid_reader gr;
// get current copy cursor into cx,cy
grid_reader_start(&gr, gr, cx, cy);
grid_reader_next_word(&gr);
grid_reader_cursor_forward(&gr);
grid_reader_get_cursor(gr, &cx, &cy);
// we now know the cx,cy is at the next word + 1 even if that is not the same line we started on
~~~~
And it would Just Work the same and take account of end-of-line, wrapping, end-of-screen and so on.

Copy mode would still need to work out what needs to be redrawn, if the selection needs to change. But I don't think it would be that difficult to do that.

**Comment by nicm** [comment]
Of course it may not be as easy as it sounds, it is hard to tell what exactly is needed because the copy mode is a bit hard to follow :-).

**Comment by topcat001** [comment]
Thanks for the ideas. I have started a little bit of exploratory coding in a new branch. I'll put up some diffs for discussion when I have something.

**Comment by topcat001** [comment]
One thing I'm trying to do is make the logic for moving the virtual cursor on the grid (assuming the API works on the grid as I have done in my branch) separate from considerations specific to copy mode such as the state of the selection, rectangle flag, etc. It takes into account information which is specific to the grid like padding cells, wrapped lines, etc. I think this will simplify the logic for the copy mode commands. The copy mode caller of the grid API will take into account copy mode state in window_copy_mode_data. Hope that makes sense.

I'll have a clearer picture once I start implementing copy mode commands on top on this API.

**Comment by nicm** [comment]
Sounds good, thanks.

**Comment by topcat001** [comment]
Hi, sorry didn't get to do too much, but here is a diff for some of the new new functions. These haven't been hooked up yet, but should be usable by the corresponding copy-mode functions. I'll continue to flesh out what's needed for a complete set, and then test implementing the copy mode functions in terms of these. Please have a look.
[git-diff.txt](https://github.com/tmux/tmux/files/5491581/git-diff.txt)

**Comment by chung1905** [comment]
I think you attached wrong file ;)

**Comment by topcat001** [comment]
Thanks! Not sure what happened! :)
[git-diff.txt](https://github.com/tmux/tmux/files/5492061/git-diff.txt)

**Comment by topcat001** [comment]
Fixed, thanks :)

**Comment by nicm** [comment]
OK this looks like a good start, I'll take a look and see if I can do any more at some point, I haven't got much time at the moment though :-/.

**Comment by nicm** [comment]
Thanks

**Comment by topcat001** [comment]
No problem, and thanks for taking a look. I still have to flesh out quite a bit more. I'm hoping that over this weekend I'll be able to implement some of the commands on top of this and test. Will keep you posted.

**Comment by nicm** [comment]
OK great stuff.

**Comment by topcat001** [comment]
This diff resolves the original issue. Much more testing is needed to make sure nothing breaks. It implements next-word and end-of-line using the new API.
[git-diff-a.txt](https://github.com/tmux/tmux/files/5539240/git-diff-a.txt)

**Comment by tuv7041** [comment]
Thanks for the patch! I just tried it, and a couple of things I notice that are different from the default bash behavior.
-It always uses the current line as a start, but the default shell goes upward if the word started before the current line (this happens with both select word and select line)
-When already on copy mode, double click stops working for me on a default tmux install. Triple click works flawlessly though.

Hope the testing helps, and thanks again for the patch, this is a great improvement over the default functionality.

**Comment by topcat001** [comment]
Thanks a lot for testing.
- Yes it's not complete yet. I need to also update the start of word and the start of line functions.
- Double click seems to work fine for me. The changes should not affect mouse behaviour.

**Comment by topcat001** [comment]
Here's an updated diff which fixes the start of word/line issue. Beware that it has only seen minimal  testing :)
[git-diff-b.txt](https://github.com/tmux/tmux/files/5541884/git-diff-b.txt)

**Comment by tuv7041** [comment]
Hey! Sorry for the delayed response. I tried the patch and its working great! A couple of small bugs I've found in the testing:
I've used the following string to test most of the functionality:

`echo "powershell -Sta -Nop -Window Hidden -EncodedCommand aQBlAHgAIAAoAE4AZQB3AC0ATwBiAGoAZQBjAHQAIABOAGUAdAAuAFcAZQBiAEMAbABpAGUAbgB0ACkALgBEAG8AdwBuAGwAbwBhAGQAUwB0AHIAaQBuAGcAKAAnAGgAdAB0AHAAOgAvAC8AMQAwAC4AMQAwAC4AMQAyADEALgAxADEAMwA6ADQANAAzAC8AbABhAHUAbgBjAGgAZQByAC4AcABzADEAJwApAA=="`

If you copy and paste this in tmux, there are a couple of differences with the raw terminal copy and paste behaviour. First, if you triple click on the base64 encoded string, the default terminal considers the double quotes character as a delimiter (it selects this instead of the full line)
`aQBlAHgAIAAoAE4AZQB3AC0ATwBiAGoAZQBjAHQAIABOAGUAdAAuAFcAZQBiAEMAbABpAGUAbgB0ACkALgBEAG8AdwBuAGwAbwBhAGQAUwB0AHIAaQBuAGcAKAAnAGgAdAB0AHAAOgAvAC8AMQAwAC4AMQAwAC4AMQAyADEALgAxADEAMwA6ADQANAAzAC8AbABhAHUAbgBjAGgAZQByAC4AcABzADEAJwApAA==`

The second bug I've found its very specific. If you use CTRL+R to reverse search that command in bash, as long as you dont end the autocomplete (as long as you see the "(reverse-i-search)`':" prompt), selecting either a word or a line works perfectly. After that, if you press any other key to complete the selection (and run the command) the selection breaks on that command. It still works perfectly for the printed string, but it won't work on the autocompleted command. Let me know if you need clarifying on this part in case I wasn't clear enough on the description. 

Everything else works perfectly! I'll keep testing but awesome patch.

**Comment by topcat001** [comment]
Great, thanks a lot! I'll work through your examples when I am able next.

**Comment by nicm** [comment]
I don't think we should treat quotes as line endings for select-line. select-word should select based on word-separators which could include quotes.

**Comment by topcat001** [comment]
The first issue should be handled through settings as @nicm suggested. I am more intrigued by the second issue.

It seems that once I run the long command, the original command string is no longer one long wrapped line in tmux's data buffer. Instead, it is broken down into a set of lines with width equal to the current pane. That is why the selection commands are behaving the way they are. This can be seen by resizing the pane width after running the command.

The output string is still one long wrapped line and everything works there. I think this has to do with how the shell is reformatting the command line. This is just my preliminary impression. I'll look more carefully into it. This might be tricky to handle, though.

**Comment by nicm** [comment]
If the shell is actually sending newlines then there is nothing we can do really, but in that case I'd expect the terminal outside tmux to also not select it as one line unless it is doing something special. If the shell is sending it as one line and tmux splits it then it could be a bug.

**Comment by topcat001** [comment]
Yes there is something funny happenign there which I don't understand yet.
BTW all my current work is in [topcat001/tmux/grid-api](/topcat001/tmux/tree/grid-api).

**Comment by nicm** [comment]
Great stuff, I'll take a look.

**Comment by topcat001** [comment]
I have had a chance to investigate the issue where the line is broken when a long command is entered and the Enter key is pressed. It looks like a bug. I'll describe my observations:
- The issue is only triggered when the cursor is moved back from the end of the command. This can be done manually by pressing the left arrow key. It also happens when a command is auto-completed.
- In this case the lines above the cursor position are not broken up. Only the lines at or below the cursor position lose their wrapped status. Also, the terminal is not doing this.
- I believe the problem is in the `screen_write_linefeed` function. When the Enter key is pressed, new content appears in the pane and the callback function is called. The call sequence is as follows:
```
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 57.1
  * frame #0: 0x000000010c014d07 tmux`screen_write_linefeed(ctx=0x00007fae8c40e950, wrapped=0, bg=8) at screen-write.c:1241:6
    frame #1: 0x000000010bfeec27 tmux`input_c0_dispatch(ictx=0x00007fae8c40e940) at input.c:1227:7
    frame #2: 0x000000010bfee919 tmux`input_parse(ictx=0x00007fae8c40e940, buf="\r\n\r\n\r\n\r\n\r\n\r\r\nll -Sta -Nop -Window Hidden -EncodedCommand aQBlAHgAIAAoAE4AZQB3AC0ATwBiAGoAZQBjAHQAIABOAGUAdAAuAFcAZQBiAEMAbABpAGUAbgB0ACkALgBEAG8AdwBuAGw...", len=13) at input.c:928:31
    frame #3: 0x000000010bfee74d tmux`input_parse_buffer(wp=0x00007faeac80a200, buf="\r\n\r\n\r\n\r\n\r\n\r\r\nll -Sta -Nop -Window Hidden -EncodedCommand aQBlAHgAIAAoAE4AZQB3AC0ATwBiAGoAZQBjAHQAIABOAGUAdAAuAFcAZQBiAEMAbABpAGUAbgB0ACkALgBEAG8AdwBuAGw...", len=13) at input.c:976:20
    frame #4: 0x000000010bfee62b tmux`input_parse_pane(wp=0x00007faeac80a200) at input.c:949:2
    frame #5: 0x000000010c060208 tmux`window_pane_read_callback(bufev=0x00007fae8c40e720, data=0x00007faeac80a200) at window.c:975:2
    frame #6: 0x000000010c0ebb47 libevent-2.1.7.dylib`bufferevent_run_readcb_ + 45
    frame #7: 0x000000010c0f0756 libevent-2.1.7.dylib`bufferevent_readcb + 334
    frame #8: 0x000000010c0f6442 libevent-2.1.7.dylib`event_process_active_single_queue + 1074
    frame #9: 0x000000010c0f3685 libevent-2.1.7.dylib`event_base_loop + 1012
    frame #10: 0x000000010c00dfdf tmux`proc_loop(tp=0x00007faeac809000, loopcb=(tmux`server_loop at server.c:245)) at proc.c:215:3
    frame #11: 0x000000010c023c97 tmux`server_start(client=0x00007faead008800, flags=268500992, base=0x00007faeac510850, lockfd=5, lockfile="") at server.c:234:2
    frame #12: 0x000000010bfa8ac5 tmux`client_connect(base=0x00007faeac510850, path="socket", flags=268500992) at client.c:162:8
    frame #13: 0x000000010bfa7e42 tmux`client_main(base=0x00007faeac510850, argc=0, argv=0x00007ffee3c5c7f0, flags=268500992, feat=0) at client.c:293:7
    frame #14: 0x000000010c0330a7 tmux`main(argc=0, argv=0x00007ffee3c5c7f0) at tmux.c:493:7
    frame #15: 0x00007fff701c2cc9 libdyld.dylib`start + 1
    frame #16: 0x00007fff701c2cc9 libdyld.dylib`start + 1
```
When the cursor is on a wrapped line, this function will reset the wrapped flag [here](https://github.com/tmux/tmux/blob/e288ea153ceffe42d88975fc23196c5dc3aeae5d/screen-write.c#L1244). One way to fix the issue is to not clear the existing flag, but I'm not sure what else will break then. Commenting it out indeed fixes the issue.

**Comment by tuv7041** [comment]
Thanks for that analysis! This debugging certainly exceeds my technical knowledge but let me know if I can help testing something (or any other way).

**Comment by nicm** [comment]
I think we can remove this and see how it goes. We are assuming that if the shell sends a linefeed then it does actually want a linebreak but I suspect few applications depend on this.

**Comment by nicm** [comment]
I took a look at your grid reader changes. I think they look good, a lot clearer.

grid.c is kind of big, how about putting them in a new file grid-reader.c or something?

What changes are left?

**Comment by topcat001** [comment]
Thanks for the updates. I was unaware of the other issues concerning newlines, but looks like all of them might get resolved by ed786309ccc7ceb2077f658e689a5f35549bb2bc.

grid-reader.c sounds good, thanks. I think I'm done with changes for now. We can merge this in. I'll have a look afterwards and think about what else can be updated to use the new functions. I'll move the grid reader stuff to the new file, clean up, and start a PR.

**Comment by topcat001** [comment]
PR https://github.com/tmux/tmux/pull/2497 submitted. Please have a look, thanks!
